XiEditor/XiTextPlane/TextLine.swift

```diff
@@ -271,5 +271,5 @@ func argbToFloats(argb: UInt32) -> (GLfloat, GLfloat, GLfloat, GLfloat) {
 func getCtLineRange(_ ctLine: CTLine, _ range: CountableRange<Int>) -> Range<GLfloat> {
     let start = GLfloat(CTLineGetOffsetForStringIndex(ctLine, range.startIndex, nil))
     let end = GLfloat(CTLineGetOffsetForStringIndex(ctLine, range.endIndex, nil))
-    return start ..< end
+    return Range(uncheckedBounds: (lower: start, upper: end))
```

Comment:
Good catch tracking it down to this. That said, I have some concerns.

* Does `uncheckedBounds` just get rid of the assert, but still violate the invariant, or is it actually valid to have a range with `upper < lower`?

* What will this do in the case of mixed ltr and rtl runs? Without digging right now, my guess would be that the signature of this function needs to be changed to return a list of ranges, and the logic needs to be updated to sense direction changes in the interior, not just sample the endpoints.

It's possible though that this is future work, and we accept some loss of accuracy for selections of bidi text for now. If my concern about the range invariant is true, then maybe we want lower to be min(start, end) and upper to be max(start, end).

Reply:
You're right, it's even nicely documented 😄
```
    /// Because this initializer does not perform any checks, it should be used
    /// as an optimization only when you are absolutely certain that `lower` is
    /// less than or equal to `upper`. Using the half-open range operator
    /// (`..<`) to form `Range` instances is preferred.
```
I updated the solution according to your suggestion.

To be honest I'm not yet sure how to approach bidi correctly. This is my first attempt to text editor development and I don't feel like I've got enough knowledge on the topic yet.